### PR TITLE
Add function:metrics and loss calculations

### PR DESCRIPTION
**Pull request by pengbo0054**

### CODE REVIEW

File: minetorch/loss.py [hunk 1]
```diff
@@ -2,146 +2,114 @@
...
-        
-        #first compute binary cross-entropy 
-        BCE = F.binary_cross_entropy(inputs, targets, reduction='mean')
-        BCE_EXP = torch.exp(-BCE)
-        focal_loss = alpha * (1-BCE_EXP)**gamma * BCE
-                       
-        return focal_loss
 
+# Below loss functions are referenced at https://www.kaggle.com/bigironsphere/loss-function-library-keras-pytorch
+def DiceLoss(inputs, targets, smooth=1):
```
Comment (louis-she) [inline minetorch/loss.py]: 方法名用下劃線， dice_loss。只有類名用駝峯命名

File: minetorch/loss.py [hunk 2]
```diff
@@ -2,146 +2,114 @@
...
+    inputs = torch.sigmoid(inputs)       
+    
+    #flatten label and prediction tensors
+    inputs = inputs.view(-1)
+    targets = targets.view(-1)
+    
+    intersection = (inputs * targets).sum()                            
+    dice_loss = 1 - (2.*intersection + smooth)/(inputs.sum() + targets.sum() + smooth)  
+    BCE = F.binary_cross_entropy(inputs, targets, reduction='mean')
+    Dice_BCE = BCE + dice_loss
```
Comment (louis-she) [inline minetorch/loss.py]: 這裏也是，只有類名會出現大寫字母

File: minetorch/loss.py [hunk 3]
```diff
@@ -2,146 +2,114 @@
...
+    
+    #comment out if your model contains a sigmoid or equivalent activation layer
+    inputs = torch.sigmoid(inputs)       
+    
+    #flatten label and prediction tensors
+    inputs = inputs.view(-1)
+    targets = targets.view(-1)
+    
+    intersection = (inputs * targets).sum()                            
+    dice_loss = 1 - (2.*intersection + smooth)/(inputs.sum() + targets.sum() + smooth)
```
Comment (louis-she) [inline minetorch/loss.py]: 4則運算符號兩邊加空格，這個加 pylint 來提示吧

File: minetorch/metrics.py [hunk 4]
```diff
@@ -1,83 +1,64 @@
...
-
-    def compute_batch(self):
-        dices = []
-        dice_cls = 0
-        if self.separate_class == True:
-             for cls in range(self.classes):
-                dice_cls = dice_single_class_batch(self.logits[:,cls], self.targets[:,cls], self.threshold)
-                dices.append(dice_cls)           
+def iou(separate_class=False):
+    @rename('iou')
```
Comment (louis-she) [inline minetorch/metrics.py]: 這裏用 `functools.partial `,
```python
def compute_iou(logits, targets, threshold=0.5, separate_class=False):
     ....

single_class_iou = functools.partial(compute_iou, separate_class=False)
multi_class_iou = functools.partial(compute_iou, separate_class=True)
```

File: minetorch/metrics.py [hunk 5]
```diff
@@ -1,83 +1,64 @@
...
+    def compute_iou(logits, targets, threshold=0.5, separate_class_=separate_class):
+        # be with the C x H x W shape
+        logits = torch.sigmoid(logits)
+        logits = logits > torch.Tensor([threshold])
+        targets = targets > torch.Tensor([0.5])
+        
+        intersection = (logits & targets).double().sum((1, 2))
```
Comment (louis-she) [inline minetorch/metrics.py]: double 改 long

File: minetorch/metrics.py [hunk 6]
```diff
@@ -1,83 +1,64 @@
...
-        if self.separate_class == True:
-             for cls in range(self.classes):
-                dice_cls = dice_single_class_batch(self.logits[:,cls], self.targets[:,cls], self.threshold)
-                dices.append(dice_cls)           
+def iou(separate_class=False):
+    @rename('iou')
+    def compute_iou(logits, targets, threshold=0.5, separate_class_=separate_class):
+        # be with the C x H x W shape
+        logits = torch.sigmoid(logits)
+        logits = logits > torch.Tensor([threshold])
```
Comment (louis-she) [inline minetorch/metrics.py]: 應該不用轉 Tensor

File: minetorch/metrics.py [hunk 7]
```diff
@@ -1,83 +1,64 @@
...
-        if self.separate_class == True:
-            for cls in range(self.classes):
-                iou_cls = iou_single_class_batch(self.logits[:,cls], self.targets[:,cls], self.threshold)
-                ious.append(iou_cls)
-        else:
-            for cls in range(self.logits.shape[1]):
-                iou_cls += iou_single_class_batch(self.logits[:,cls], self.targets[:,cls], self.threshold)
-            ious.append(iou_cls/self.classes)
+def dice(separate_class=False):
+    @rename('dice')
```
Comment (louis-she) [inline minetorch/metrics.py]: 同上，用 `partial`

File: minetorch/trainer.py [hunk 8]
```diff
@@ -264,25 +265,33 @@ def train(self):
...
             self.notebook_output(f'start to train epoch {self.current_epoch}')
             for index, data in enumerate(self.tqdm(self.train_dataloader)):
                 if self.trival is True and index == 10:
                     break
-                train_loss, train_metric = self.run_train_iteration(index, data, train_iters)
+                train_loss, train_metrics = self.run_train_iteration(index, data, train_iters)
                 total_train_loss += train_loss
-                total_train_metrics.append(train_metric)
+                for metric in train_metrics:
+                    if metric not in total_train_metrics.keys():
```
Comment (louis-she) [inline minetorch/trainer.py]: 不需要加 `.keys`

File: minetorch/trainer.py [hunk 9]
```diff
@@ -264,25 +265,33 @@ def train(self):
...
+                        total_train_metrics[metric] = 0
+                        total_train_metrics[metric] += train_metrics[metric]
+                    else:
+                        total_train_metrics[metric] += train_metrics[metric]
```
Comment (louis-she) [inline minetorch/trainer.py]: 這裏這麼寫，
```python
if metric not in total_train_metrics:
    total_train_metrics[metric] = 0
total_train_metrics[metric] += train_metrics[metric]
```
可以幹掉 `else` ，不然有兩行相同的代碼
